source/common/filter/http/filter_config_discovery_impl.cc

```diff
@@ -196,24 +200,65 @@ std::shared_ptr<FilterConfigSubscription> FilterConfigProviderManagerImpl::getSu
 }
 
 FilterConfigProviderPtr FilterConfigProviderManagerImpl::createDynamicFilterConfigProvider(
-    const envoy::config::core::v3::ConfigSource& config_source,
-    const std::string& filter_config_name, const std::set<std::string>& require_type_urls,
-    Server::Configuration::FactoryContext& factory_context, const std::string& stat_prefix,
-    bool apply_without_warming) {
-  auto subscription =
-      getSubscription(config_source, filter_config_name, factory_context, stat_prefix);
+    const envoy::config::core::v3::ExtensionConfigSource& config_source,
+    const std::string& filter_config_name, Server::Configuration::FactoryContext& factory_context,
+    const std::string& stat_prefix) {
+  auto subscription = getSubscription(config_source.config_source(), filter_config_name,
+                                      factory_context, stat_prefix);
   // For warming, wait until the subscription receives the first response to indicate readiness.
   // Otherwise, mark ready immediately and start the subscription on initialization. A default
   // config is expected in the latter case.
-  if (!apply_without_warming) {
+  if (!config_source.apply_default_config_without_warming()) {
     factory_context.initManager().add(subscription->initTarget());
   }
-  auto provider = std::make_unique<DynamicFilterConfigProviderImpl>(
-      std::move(subscription), require_type_urls, factory_context);
+  absl::flat_hash_set<std::string> require_type_urls;
+  for (const auto& type_url : config_source.type_urls()) {
+    auto factory_type_url = TypeUtil::typeUrlToDescriptorFullName(type_url);
+    require_type_urls.emplace(factory_type_url);
+  }
+  auto provider = std::make_unique<DynamicFilterConfigProviderImpl>(subscription, require_type_urls,
+                                                                    factory_context);
   // Ensure the subscription starts if it has not already.
-  if (apply_without_warming) {
+  if (config_source.apply_default_config_without_warming()) {
     factory_context.initManager().add(provider->init_target_);
   }
+
+  // If the subscription already received a config, attempt to apply it.
+  bool last_config_valid = false;
+  if (subscription->lastConfig().has_value()) {
+    try {
+      provider->validateTypeUrl(subscription->lastTypeUrl());
```

Comment:
Would there be a value in moving this to returning true/false?

Reply:
I considered it, but the exception contains the list of required type URLs, which is useful. Bool would be less actionable.